app/src/main/java/org/luxoft/sdl_core/BluetoothLongWriter.java

```diff
@@ -30,9 +33,14 @@ public void setCallback(LongWriterCallback callback) {
         mCallback = callback;
     }
 
-    private byte[] prepareMessageToSend(final int frames_left, final byte[] message) {
+    private byte[] prepareMessageToSend(final int frames_left, final byte[] message, final boolean compression_flag) {
         ByteBuffer buffer = ByteBuffer.allocate(message.length + mFramesCountBytes);
-        buffer.putInt(frames_left);
+
+        final int compress_bit = 0x40000000;
+
+        int t_frames_left = compression_flag ? frames_left | compress_bit : frames_left;
```

Comment:
@VadymLuchko well, it's a tricky way to write and read information from a message.
I would suggest keeping the compression flag as a 5th bit.
Then in the long writer, you can simply use the `put()` method and put 1 or 0 as a byte.
Also, rename `mFramesCountBytes` -> `mHeaderSize` and set it to 5
In long reader, you can simply use `get()` method twice in a row:
```
private int extractFramesCount(byte[] value) {
        byte[] frames = Arrays.copyOfRange(value, 0, mMessageOffset);
        ByteBuffer buffer = ByteBuffer.wrap(frames);
        final int frames_count = buffer.getInt();
        final boolean is_compressed = buffer.get() == 1; // or something similar
        return ...
    }
```

Also, please rename `mMessageOffset` -> `mHeaderSize` and set it to 5
In this approach, it is more obvious and you don't need to use binary operations in your code

Reply:
@AKalinich-Luxoft 
Hope I understood correctly and here it meant the 5th byte and not bit )
fixed in 03157c3

Reply:
@VadymLuchko right, nice job :)